### ISSUE TITLE
The relesae of full code

### ISSUE THREAD

**Issue by honghd16** [Feliciaxyao/ICML2024-FSTTA#1]
Hi Xuan,

Thank you for sharing your impressive work. I am particularly interested in how you've integrated TTA into VLN. I noticed that the current version of the code is missing some critical components, such as fstta.py in the models directory, which prevents it from running. Do you have any plans to release the complete codebase?

Cheers

**Comment by Feliciaxyao** [comment]
Thank you for your interest! The key related code for FSTTA can be found in the /map_nav_src/models directory. If you have any further questions, feel free to reach out!

**Comment by honghd16** [comment]
Thanks for sharing the code!
When I ran this code, I found several issues with the variable definition.
Including the id, mean_loss, ref_diff in the agent_obj.py:
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/4582ed0abf1b890079c6f1acfd2d5488023b2115/map_nav_src/reverie/agent_obj.py#L568
Also, the FSTTA.py also has some problems with the fast_forward_and_adapt function:
```
# "Gradient Accumulation"
    grad_stack += loss.grad
    m_loss += loss
    params_stack = torch.stack(loss.grad)
```
These three variables are not initialized and only as local variables, which will not be persistent through three steps.
I am not sure what the loss.grad means here because it always gives back the NoneType.
Any help is appreciated!

**Comment by Feliciaxyao** [comment]
Hi, sorry for my late answer. I’ve reviewed the issues you mentioned. During the fast adaptation stage, we first perform gradient collection for parameters in the updated layer normalization layers, as shown here:
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/b9dc2940ba85d92f173909f6c36afdebd912f3c1/map_nav_src/models/FSTTA.py#L145-L150

Following this, gradient accumulation is performed on the parameters themselves, not the loss, which can be seen here:
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/b9dc2940ba85d92f173909f6c36afdebd912f3c1/map_nav_src/models/FSTTA.py#L186-L191

**Comment by honghd16** [comment]
Thanks for the response. I tried the updated code but found that since the fast adaptation is performed every M steps, the FAST object is initialized for each step:
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/b9dc2940ba85d92f173909f6c36afdebd912f3c1/map_nav_src/reverie/agent_obj.py#L421
However, in the fstta.py, it only uses a local variable grad_stack to store the gradients :
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/b9dc2940ba85d92f173909f6c36afdebd912f3c1/map_nav_src/models/FSTTA.py#L154
This will only take effect at the M-th step because the previous M-1 gradients have been cleared, causing the if-condition in this line never be satisfied because the len(grads) is always 1:
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/b9dc2940ba85d92f173909f6c36afdebd912f3c1/map_nav_src/models/FSTTA.py#L187
which will finally cause this problem:
![image](https://github.com/user-attachments/assets/f9ca092a-8f03-44d4-aafb-12d935dcbd6f)

Besides this, I also have a problem with the slow update process. Since the slow update is performed in every N sample, the code use the variable id to deal with the sample number:
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/b9dc2940ba85d92f173909f6c36afdebd912f3c1/map_nav_src/reverie/agent_obj.py#L557
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/b9dc2940ba85d92f173909f6c36afdebd912f3c1/map_nav_src/reverie/agent_obj.py#L571-L574
However, the id is only a local variable in this rollout function, and I could not find the definition for it. It is also confusing to understand this code:
https://github.com/Feliciaxyao/ICML2024-FSTTA/blob/b9dc2940ba85d92f173909f6c36afdebd912f3c1/map_nav_src/reverie/agent_obj.py#L561-L568
Since all the variables like params_stack_slow and ref-diff seem to include multiple sample information, but only as variables for one sample here.

**Comment by Anonymous-N** [comment]
Hello,

Thank you for your great work! I have noticed an issue while using your open-source code. As mentioned, the `rollout` function contains multiple undefined local variables, including `id`, `ref_diff`, and `loss_sum`. If you plan to make updates, I would love to know the timeline. Looking forward to your response!

Thank you!
